llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp

```diff
@@ -3882,7 +3882,15 @@ void SelectionDAGBuilder::visitUIToFP(const User &I) {
   SDValue N = getValue(I.getOperand(0));
   EVT DestVT = DAG.getTargetLoweringInfo().getValueType(DAG.getDataLayout(),
                                                         I.getType());
-  setValue(&I, DAG.getNode(ISD::UINT_TO_FP, getCurSDLoc(), DestVT, N));
+  bool IsNonNeg = false;
+  if (auto *PNI = dyn_cast<PossiblyNonNegInst>(&I))
+    IsNonNeg = true;
+
+  unsigned Opc = ISD::UINT_TO_FP;
+  if (IsNonNeg && DAG.getTargetLoweringInfo().isSIToFPCheaperThanUIToFP(
```

Comment:
Do you need to do the optimization in the builder or leave it to the combiner? The IRBuilder blindly copies flags:
https://github.com/llvm/llvm-project/blob/22e7e68a40b8b1aac8b44137685d21ac4b98bd17/llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp#L300

Reply:
It should have been unary op.

Reply:
To translate and leave to combiner need `nneg` support on `uitofp` in the DAG. I figured this accomplished the job and was simpler. I don't know if we have any issues of generating wrong-signed casts in the backend.

Reply:
The combiner could learn the nneg flag, then we have the same optimization twice.

Reply:
Okay, Ill update.